src/pkg/task/execution.go

```diff
@@ -43,3 +53,211 @@ type ExecutionManager interface {
 	// List executions according to the query
 	List(ctx context.Context, query *q.Query) (executions []*Execution, err error)
 }
+
+// NewExecutionManager return an instance of the default execution manager
+func NewExecutionManager() ExecutionManager {
+	return &executionManager{
+		executionDAO: dao.NewExecutionDAO(),
+		taskMgr:      Mgr,
+		taskDAO:      dao.NewTaskDAO(),
+	}
+}
+
+type executionManager struct {
+	executionDAO dao.ExecutionDAO
+	taskMgr      Manager
+	taskDAO      dao.TaskDAO
+}
+
+func (e *executionManager) Create(ctx context.Context, vendorType string, vendorID int64, trigger string,
+	extraAttrs ...map[string]interface{}) (int64, error) {
+	extras := map[string]interface{}{}
+	if len(extraAttrs) > 0 && extraAttrs[0] != nil {
+		extras = extraAttrs[0]
+	}
+	data, err := json.Marshal(extras)
+	if err != nil {
+		return 0, err
+	}
+
+	execution := &dao.Execution{
+		VendorType: vendorType,
+		VendorID:   vendorID,
+		Trigger:    trigger,
+		ExtraAttrs: string(data),
+		StartTime:  time.Now(),
+	}
+	return e.executionDAO.Create(ctx, execution)
+}
+
+func (e *executionManager) UpdateStatus(ctx context.Context, id int64, status, message string, endTime time.Time) error {
+	return e.executionDAO.Update(ctx, &dao.Execution{
+		ID:            id,
+		Status:        status,
+		StatusMessage: message,
+		EndTime:       endTime,
+	}, "Status", "StatusMessage", "EndTime")
+}
+
+func (e *executionManager) Stop(ctx context.Context, id int64) error {
+	tasks, err := e.taskDAO.List(ctx, &q.Query{
+		Keywords: map[string]interface{}{
+			"ExecutionID": id,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	for _, task := range tasks {
+		if err = e.taskMgr.Stop(ctx, task.ID); err != nil {
+			log.Errorf("failed to stop task %d: %v", task.ID, err)
+			continue
+		}
+	}
+	return nil
+}
+
+func (e *executionManager) Delete(ctx context.Context, id int64) error {
+	tasks, err := e.taskDAO.List(ctx, &q.Query{
+		Keywords: map[string]interface{}{
+			"ExecutionID": id,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, task := range tasks {
+		if !job.Status(task.Status).Final() {
```

Comment:
Is it possible to treat this operation as a default one? That means if the task is not completed, the deleting action will try to stop it first.

Reply:
A corner case: think about an execution which contains two tasks `a` and `b`, both of them are running. Now I try to delete the execution, stop `a` successfully, delete `a` from database, try to stop b but get error, as the operation is in a DB transaction, all database deletion rolls back, this causes the inconsistent between database and jobservice.

Reply:
ok, let's keep it with a simple way.